File: android/app/src/test/java/net/crimsys/app/domain/legal/EgyptianDeadlineCalculatorTest.kt
package net.crimsys.app.domain.legal

import java.time.LocalDate
import net.crimsys.app.domain.legal.EgyptianDeadlineCalculator.DeadlineResult
import net.crimsys.app.domain.legal.EgyptianDeadlineCalculator.LegalDeadlineChannel
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Unit tests for the M4 dual-verification deadline calculator.
 *
 * The `now`-dependent field ([DeadlineResult.Verified.daysRemaining]) is
 * deliberately not asserted — everything else (raw deadline, weekend roll,
 * verification status) is deterministic for fixed start dates.
 *
 * The Engine-2 arithmetic itself is additionally cross-verified against a
 * `java.time`-independent reference over 1990–2060 (see repo docs); these
 * tests pin the observable contract.
 */
class EgyptianDeadlineCalculatorTest {

    private val calc = EgyptianDeadlineCalculator()

    // ------------------------------------------------------------------
    // Channel metadata — citation integrity policy
    // ------------------------------------------------------------------
    @Test
    fun `all channels are flagged as needing legal review`() {
        // A wrong legal date is a catastrophe; the badge must ship ON until
        // counsel signs off. If this test fails, counsel signed off — update
        // the flag AND the cited articles together.
        LegalDeadlineChannel.entries.forEach { channel ->
            assertTrue(channel.needsLegalReview)
        }
    }

    @Test
    fun `channel lookup by id round-trips`() {
        LegalDeadlineChannel.entries.forEach { channel ->
            assertEquals(channel, LegalDeadlineChannel.fromId(channel.id))
        }
        assertNull(LegalDeadlineChannel.fromId("nope"))
    }

    // ------------------------------------------------------------------
    // Weekend roll (Fri/Sat → Sunday), verified against reference dates
    // ------------------------------------------------------------------
    @Test
    fun `opposition window landing on Friday rolls to Sunday`() {
        // 2026-09-01 + 10d = 2026-09-11 (Friday) → rolls to 2026-09-13.
        val r = calc.compute(LegalDeadlineChannel.OPPOSITION, LocalDate.of(2026, 9, 1))
        assertTrue(r is DeadlineResult.Verified)
        r as DeadlineResult.Verified
        assertEquals(LocalDate.of(2026, 9, 11), r.rawDeadline)
        assertEquals(LocalDate.of(2026, 9, 13), r.deadline)
        assertTrue(r.wasRolled)
    }

    @Test
    fun `cassation window landing on a weekday is not rolled`() {
        // 2026-09-03 + 60d = 2026-11-02 (Monday) → unchanged.
        val r = calc.compute(LegalDeadlineChannel.CASSATION, LocalDate.of(2026, 9, 3))
        assertTrue(r is DeadlineResult.Verified)
        r as DeadlineResult.Verified
        assertEquals(LocalDate.of(2026, 11, 2), r.deadline)
        assertFalse(r.wasRolled)
    }

    @Test
    fun `roll across a month boundary lands on the first of next month`() {
        // 2026-10-21 + 10d = 2026-10-31 (Saturday) → rolls to 2026-11-01.
        val r = calc.compute(LegalDeadlineChannel.OPPOSITION, LocalDate.of(2026, 10, 21))
        assertTrue(r is DeadlineResult.Verified)
        r as DeadlineResult.Verified
        assertEquals(LocalDate.of(2026, 10, 31), r.rawDeadline)
        assertEquals(LocalDate.of(2026, 11, 1), r.deadline)
        assertTrue(r.wasRolled)
    }

    @Test
    fun `deadline on Saturday rolls exactly one day`() {
        // 2026-10-28 + 10d = 2026-11-07 (Saturday) → rolls to 2026-11-08.
        val r = calc.compute(LegalDeadlineChannel.OPPOSITION, LocalDate.of(2026, 10, 28))
        assertTrue(r is DeadlineResult.Verified)
        r as DeadlineResult.Verified
        assertEquals(LocalDate.of(2026, 11, 8), r.deadline)
        assertTrue(r.wasRolled)
    }

    // ------------------------------------------------------------------
    // Leap-year arithmetic through both engines
    // ------------------------------------------------------------------
    @Test
    fun `window ending on a leap day is computed consistently`() {
        // 2028-02-19 + 10d = 2028-02-29 (Tuesday) → no roll needed.
        val r = calc.compute(LegalDeadlineChannel.OPPOSITION, LocalDate.of(2028, 2, 19))
        assertTrue(r is DeadlineResult.Verified)
        r as DeadlineResult.Verified
        assertEquals(LocalDate.of(2028, 2, 29), r.deadline)
        assertFalse(r.wasRolled)
    }

    // ------------------------------------------------------------------
    // Input validation
    // ------------------------------------------------------------------
    @Test
    fun `future start date is rejected`() {
        val r = calc.compute(LegalDeadlineChannel.CASSATION, LocalDate.now().plusDays(1))
        assertTrue(r is DeadlineResult.Invalid)
    }

    // ------------------------------------------------------------------
    // Dual verification contract
    // ------------------------------------------------------------------
    @Test
    fun `both engines agree on a sweep of start dates`() {
        // Every result must be Verified — a sweep across month and leap
        // boundaries exercises both independent implementations.
        val starts = listOf(
            LocalDate.of(2025, 12, 20), LocalDate.of(2026, 2, 20),
            LocalDate.of(2026, 9, 1), LocalDate.of(2027, 12, 25),
            LocalDate.of(2028, 2, 19), LocalDate.of(2028, 12, 30),
        )
        starts.forEach { start ->
            LegalDeadlineChannel.entries.forEach { channel ->
                val r = calc.compute(channel, start)
                assertTrue("engine disagreement at $start/$channel", r is DeadlineResult.Verified)
            }
        }
    }
}

File: mobile_app/test/egyptian_deadline_calculator_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:legal_clay/domain/legal/egyptian_deadline_calculator.dart';

/// Unit tests for the dual-verification deadline calculator — the Dart port
/// of `EgyptianDeadlineCalculatorTest.kt`. All dates are pinned so the suite
/// is deterministic; `daysRemaining` is the only now-derived field.
void main() {
  const calc = EgyptianDeadlineCalculator();
  final today = DateTime.utc(2026, 1, 1);

  // ------------------------------------------------------------------
  // Channel metadata — citation integrity policy
  // ------------------------------------------------------------------
  test('all channels are flagged as needing legal review', () {
    // A wrong legal date is a catastrophe; the badge must ship ON until
    // counsel signs off. If this test fails, counsel signed off — update
    // the flag AND the cited articles together.
    for (final channel in LegalDeadlineChannel.values) {
      expect(channel.needsLegalReview, isTrue);
    }
  });

  test('channel lookup by id round-trips', () {
    for (final channel in LegalDeadlineChannel.values) {
      expect(LegalDeadlineChannel.fromId(channel.id), channel);
    }
    expect(LegalDeadlineChannel.fromId('nope'), isNull);
  });

  // ------------------------------------------------------------------
  // Weekend roll (Fri/Sat → Sunday), verified against reference dates
  // ------------------------------------------------------------------
  test('opposition window landing on Friday rolls to Sunday', () {
    // 2026-09-01 + 10d = 2026-09-11 (Friday) → rolls to 2026-09-13.
    final r = calc.compute(
      LegalDeadlineChannel.opposition,
      DateTime.utc(2026, 9, 1),
      today: today,
    );
    expect(r, isA<DeadlineVerified>());
    final v = r as DeadlineVerified;
    expect(v.rawDeadline, DateTime.utc(2026, 9, 11));
    expect(v.deadline, DateTime.utc(2026, 9, 13));
    expect(v.wasRolled, isTrue);
  });

  test('cassation window landing on a weekday is not rolled', () {
    // 2026-09-03 + 60d = 2026-11-02 (Monday) → unchanged.
    final r = calc.compute(
      LegalDeadlineChannel.cassation,
      DateTime.utc(2026, 9, 3),
      today: today,
    );
    expect(r, isA<DeadlineVerified>());
    final v = r as DeadlineVerified;
    expect(v.deadline, DateTime.utc(2026, 11, 2));
    expect(v.wasRolled, isFalse);
  });

  test('roll across a month boundary lands on the first of next month', () {
    // 2026-10-21 + 10d = 2026-10-31 (Saturday) → rolls to 2026-11-01.
    final r = calc.compute(
      LegalDeadlineChannel.opposition,
      DateTime.utc(2026, 10, 21),
      today: today,
    );
    expect(r, isA<DeadlineVerified>());
    final v = r as DeadlineVerified;
    expect(v.rawDeadline, DateTime.utc(2026, 10, 31));
    expect(v.deadline, DateTime.utc(2026, 11, 1));
    expect(v.wasRolled, isTrue);
  });

  test('deadline on Saturday rolls exactly one day', () {
    // 2026-10-28 + 10d = 2026-11-07 (Saturday) → rolls to 2026-11-08.
    final r = calc.compute(
      LegalDeadlineChannel.opposition,
      DateTime.utc(2026, 10, 28),
      today: today,
    );
    expect(r, isA<DeadlineVerified>());
    final v = r as DeadlineVerified;
    expect(v.deadline, DateTime.utc(2026, 11, 8));
    expect(v.wasRolled, isTrue);
  });

  // ------------------------------------------------------------------
  // Leap-year arithmetic through both engines
  // ------------------------------------------------------------------
  test('window ending on a leap day is computed consistently', () {
    // 2028-02-19 + 10d = 2028-02-29 (Tuesday) → no roll needed.
    final r = calc.compute(
      LegalDeadlineChannel.opposition,
      DateTime.utc(2028, 2, 19),
      today: today,
    );
    expect(r, isA<DeadlineVerified>());
    final v = r as DeadlineVerified;
    expect(v.deadline, DateTime.utc(2028, 2, 29));
    expect(v.wasRolled, isFalse);
  });

  // ------------------------------------------------------------------
  // Input validation
  // ------------------------------------------------------------------
  test('future start date is rejected', () {
    final r = calc.compute(
      LegalDeadlineChannel.cassation,
      DateTime.utc(2026, 1, 2),
      today: today,
    );
    expect(r, isA<DeadlineInvalid>());
  });

  // ------------------------------------------------------------------
  // Dual verification contract
  // ------------------------------------------------------------------
  test('both engines agree on a sweep of start dates', () {
    // Every result must be Verified — a sweep across month and leap
    // boundaries exercises both independent implementations.
    const starts = [
      (2025, 12, 20), (2026, 2, 20), (2026, 9, 1),
      (2027, 12, 25), (2028, 2, 19), (2028, 12, 30),
    ];
    for (final (y, m, d) in starts) {
      for (final channel in LegalDeadlineChannel.values) {
        final r = calc.compute(
          channel,
          DateTime.utc(y, m, d),
          today: today,
        );
        expect(r, isA<DeadlineVerified>(),
            reason: 'engine disagreement at $y-$m-$d/${channel.id}');
      }
    }
  });

  test('engine-2 YMD mirrors the DateTime deadline exactly', () {
    final r = calc.compute(
      LegalDeadlineChannel.cassation,
      DateTime.utc(2026, 9, 3),
      today: today,
    );
    final v = r as DeadlineVerified;
    expect(v.deadlineYMD.year, v.deadline.year);
    expect(v.deadlineYMD.month, v.deadline.month);
    expect(v.deadlineYMD.day, v.deadline.day);
    expect(v.rawDeadlineYMD.year, v.rawDeadline.year);
    expect(v.rawDeadlineYMD.month, v.rawDeadline.month);
    expect(v.rawDeadlineYMD.day, v.rawDeadline.day);
  });

  test('daysRemaining is measured from the anchored today', () {
    final r = calc.compute(
      LegalDeadlineChannel.opposition,
      DateTime.utc(2026, 9, 1),
      today: DateTime.utc(2026, 9, 5),
    );
    final v = r as DeadlineVerified;
    // 2026-09-05 → deadline 2026-09-13 = 8 days.
    expect(v.daysRemaining, 8);
  });
}
